Adding support for external C functions that have integer (or empty) args and/or returns

emarteca opened the pull request:
Starts addressing @https://github.com/rust-lang/miri/issues/11

### Implementation
Adding support for calling external C functions that have any number of integer arguments (types of integers: `i8`, `i16`, `i32`, `i64`, `u8`, `u16`, `u32`, `u64`) and an integer return type (or `void`).
As suggested in @https://github.com/rust-lang/miri/issues/11, the [`libffi` crate](https://docs.rs/libffi/latest/libffi/index.html) is used to dispatch the calls to external C functions.

#### Modifications
Main modifications are to:
* [helper.rs](https://github.com/emarteca/miri/blob/int-function-args-returns/src/helpers.rs) : adding a function `call_and_add_external_c_fct_to_context` to read the code pointer to the external C function, dispatch the call, and save the return in MIRI internal memory. Handles all conversions between MIRI and C values (using some macros, also defined in this file). 
* [foreign_items.rs](https://github.com/emarteca/miri/blob/int-function-args-returns/src/shims/foreign_items.rs) : handles the calling of `call_and_add_external_c_fct_to_context` in [helper.rs](https://github.com/emarteca/miri/blob/int-function-args-returns/src/helpers.rs) when a foreign item is encountered. Also adds some structs to model C representations of arguments, and the signature of the external C call.

### Testing
Adds tests for the following external functions which are now supported:
* [int tests](https://github.com/emarteca/miri/blob/int-function-args-returns/tests/pass/external_C/int_c_tests.rs): 
     - adds 2 to a provided int (no type of int specified, so autocasts)
     - takes the sum of its 12 arguments (tests stack spill)
     - adds 3 to a 16 bit int
     - adds an `i16` to an `i64`
     - returns -10 as an unsigned int
* [void tests](https://github.com/emarteca/miri/blob/int-function-args-returns/tests/pass/external_C/print_from_c.rs)
     - void function that prints from C
  
### Code review
The code in this PR was reviewed by @maurer on [another fork](https://github.com/maurer/miri/pull/1) -- thanks!

LegNeato commented:
This is awesome! It would be nice to have this support as optional / gated...similar to how miri has an option for isolation, it is useful to assert that no c code is called (if you take the extreme view that C = UB 😄 ) .

oli-obk commented on src/eval.rs:
```diff
@@ -126,6 +127,9 @@ pub struct MiriConfig {
     /// Whether Stacked Borrows retagging should recurse into fields of datatypes.
     pub retag_fields: bool,
+    /// the location of a shared object file to load when calling external C functions
+    /// TODO! this is probably a temp solution
```
Elaborate here or link to the plan for replacing this. (or remove the TODO ^^)

oli-obk commented on src/helpers.rs:
```diff
@@ -69,6 +75,54 @@ fn try_resolve_did<'tcx>(tcx: TyCtxt<'tcx>, path: &[&str]) -> Option<DefId> {
     )
 }
 
+/// if the primitive provided can be converted to a type matching the hir type pattern
+/// prim_type_pat (conversion through prim_conv_fct), then create a CArg of this primitive
+/// value with the carg_constructor provided
```
except for the `return` part, all of this logic could be regular Rust functions. I think I'd prefer that, and then calling the `return` manually.

The `prim_type_pat` can just be matched by equality instead of destructuring for this to work.

oli-obk commented on src/helpers.rs:
```diff
@@ -69,6 +75,54 @@ fn try_resolve_did<'tcx>(tcx: TyCtxt<'tcx>, path: &[&str]) -> Option<DefId> {
+    };
+}
+
+/// if the return type (ret_type) of a function is a primitive integer type pattern in hir (int_type_pat),
+/// then call the function (ptr) with arguments libffi_args, store the return value as the specified
+/// primitive integer type (int_type), and then write this value out to the miri memory as an integer
+macro_rules! call_and_store_int_return {
```
Same thing here, if a macro can be avoided, I think that would be better.

oli-obk commented on src/helpers.rs:
```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
+
+    // extract the scalar value from the result of reading a scalar from the machine
+    // and convert it to a CArg
```
General notes on documentation and comments:

* Prefer doc comments over regular comments (if the site allows doc comments, which this one does)
* Start with a capitalized letter
* Put things like `CArg` or other names into backticks
* End sentences with a `.`

oli-obk commented on src/helpers.rs:
```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+    // extract the scalar value from the result of reading a scalar from the machine
+    // and convert it to a CArg
+    fn extract_scalar(
+        scalar_res: InterpResult<'tcx, ScalarMaybeUninit<Tag>>,
```
I think you should pass in just a `ScalarMaybeUninit` and `?` away the `InterpResult` at the call sites. The unsupported argument message is misleading imo.

emarteca replied:
good idea

oli-obk commented on src/helpers.rs:
```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        libffi_args: Vec<libffi::high::Arg<'a>>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_mut();
+        unsafe {
```
style nit: only wrap the actually unsafe part in `unsafe {}` (and document why the unsafe operation inside is sound at this site).

oli-obk commented on src/helpers.rs:
```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        let this = self.eval_context_mut();
+        unsafe {
+            // ints
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I8), i8, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I16), i16, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I32), i32, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I64), i64, dest, libffi_args, ptr, external_fct_defn.output_type);
+            // uints
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U8), u8, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U16), u16, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U32), u32, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U64), u64, dest, libffi_args, ptr, external_fct_defn.output_type);
```
Instead of trying for each type, inline the macro here, match the `PrimTy` once and dispatch from there.

oli-obk commented on src/machine.rs:
```diff
@@ -398,6 +401,15 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
             since_progress_report: 0,
+            external_c_so_file: match &config.external_c_so_file {
+                Some(lib_file_path) =>
+                    Some(unsafe {
```
document that this is safe in the sense that it's the caller's obligation to provide a correct shared object

oli-obk commented on src/shims/foreign_items.rs:
```diff
@@ -359,6 +441,21 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                &tcx.hir().get(tcx.hir().local_def_id_to_hir_id(local_id)),
+                link_name,
+            ) {
+                if let Ok(_) =
+                    this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)
```
The errors should be forwarded with `?` and this entire code here should be gated on there being a `external_c_so_file` specified.

oli-obk reviewed (changes requested):
This is really cool!

RalfJung commented:
> This is awesome! It would be nice to have this support as optional / gated...similar to how miri has an option for isolation, it is useful to assert that no c code is called (if you take the extreme view that C = UB smile ) .

Oh definitely, this should be off-by-default.

emarteca commented:
Hey guys, thanks for the quick and thorough review! And hello fellow Boston squad :)
I'll go through these changes and push an updated version of the code soon.

emarteca commented on src/helpers.rs:
```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+
+        // call the C function using libffi
+        unsafe {
+            let lib = match this.machine.external_c_so_file.as_ref() {
+                Some(x) => x,
+                None => {
+                    throw_unsup_format!("No shared object file specified");
+                }
+            };
```
good point

emarteca commented:
Hey all -- thanks again for the feedback! I've just pushed an updated version of the code, with the main changes being:
* removed macros: instead [now](https://github.com/emarteca/miri/blob/10c344bea21eb8bc18919e45c193ae54ca7895cb/src/c_ffi_support.rs#L15) just matching on the pattern and dispatching directly (and similarly for the [other macro](https://github.com/emarteca/miri/blob/10c344bea21eb8bc18919e45c193ae54ca7895cb/src/c_ffi_support.rs#L80))
* modified some of the error handling logic in a few cases as requested by @oli-obk 
* moved all the C FFI supporting functionality into its own file [`c_ffi_support.rs`](https://github.com/emarteca/miri/blob/10c344bea21eb8bc18919e45c193ae54ca7895cb/src/c_ffi_support.rs) as requested by @RalfJung 
* updated comment style as requested by @oli-obk 

I did not fix Oli's comment about [supporting the external C tests only on certain platforms](https://github.com/rust-lang/miri/pull/2363/files/adfc37181847e245e30d3fee6ef13b5693ca8896#diff-1bf404991ea5fd49d3a0221d9cf055dedf186f899f33610263a403bb745bd746) as I'm not sure how best to do this.

oli-obk commented on tests/compiletest.rs:
```diff
@@ -38,6 +38,9 @@ fn run_tests(mode: Mode, path: &str, target: Option<String>) {
         flags.push(target.clone());
     }
 
+    // add flags for testing the external C functionality
+    flags.push("-Zmiri-external_c_so_file=tests/external_C/libtestlib.so".to_string());
```
I think we should add a test to see what happens if the flag is used with missing/wrong files, and if the diagnostics are reasonable, just stick with that.

Which reminds me: this doesn't need to go here. You can add it to just the tests that actually use it by adding a `//@compile-flags: -Zmiri....` comment to them. This will also make filtering the tests for supported platforms easier, as you can now use `//@only-linux` or sth to make sure the tests are only run there

oli-obk commented on src/machine.rs:
```diff
@@ -398,6 +401,17 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
             preemption_rate: config.preemption_rate,
             report_progress: config.report_progress,
             since_progress_report: 0,
+            external_c_so_lib: match &config.external_c_so_file {
```
use `map` instead of `match` here.

emarteca replied:
nice! that's cleaner

oli-obk commented on src/machine.rs:
```diff
@@ -398,6 +401,17 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
+                // Note: it is the user's responsibility to provide a correct SO file
+                // and if this condition is met, then this library loading is a safe operation.
+                    Some(unsafe {
+                        libloading::Library::new(lib_file_path).unwrap_or_else(|_| {
+                            panic!("Error reading specified shared object file")
+                        })
```
```suggestion
                        libloading::Library::new(lib_file_path).expect("Failed to read specified shared object file")
```

oli-obk commented on src/shims/foreign_items.rs:
```diff
@@ -359,6 +363,27 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    // An error here means that the function being called was not successfully called
+                    // via the specified SO file; we should continue and check if it corresponds to
+                    // a provided shim.
+                    if let Ok(_) =
+                        this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)
+                    {
+                        return Ok(EmulateByNameResult::ExecutedExternalCCall);
+                    }
```
I think we should error out if this failed. The only "soft failure" mode I'd expect would be that the shared object does not export this function. You could make `call_and_add_external_c_fct_to_context` return a `InterpResult<bool>` that is only Ok(`true`) if a function has been found (all failures should be an `Err`). Then you can check the boolean here
```suggestion
                    if this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)?
                    {
                        return Ok(EmulateByNameResult::ExecutedExternalCCall);
                    }
```

emarteca replied:
Hmm, ok I see your point; and you're right that the main error case is when the SO doesn't export the function. Right now the issue is that since we're trying to call the function through the SO before we try the shims, if we error out for a function not being found in the SO then the shims will never be tried. There's 2 potential solutions I see:
1. don't error out if the function is not exported by the SO and continue to the shims execution 
2. try the shims before we try the function via the SO

For now I'll implement 1 but I'm happy to switch :)

RalfJung replied:
I have not read all of the discussion yet, but I do think the shims we have should usually take precedence. Or is there a case where having the FFI take precedence is important?

oli-obk replied:
Probably when we have a shim returning a file handle and then that handle goes to a FFI function. That will go horribly wrong

oli-obk replied:
I thought about solution 1 when I wrote my comment. It replicates what you have right now with little changes. I just don't want to hide the argument type mismatch errors, which your current logic does by "catching" errors here

RalfJung replied:
Yeah we have that file handle issue both ways though -- if a function is missing from the SO and we use the shim instead, things will still go horribly wrong.

emarteca replied:
One small issue with both of these solutions that I'm currently running into is that the failing test output is now different when a SO is provided and the function being called is not found (since the error will say something like "function not exported in SO"). With my old code this didn't come up because with any error in using the SO would go forward and try with the shims, and if it still failed there the error message would be the same as it was originally.

RalfJung replied:
If the idea is that the SO *replaces* all shims, then that seems fine?

Don't run all the existing tests with the SO, that makes little sense then. Instead have a few tests dedicated at calling functions in the SO. The existing tests are for core Miri features and the Miri shims, after all, I wouldn't expect the SO to reproduce them -- Miri shims will always be better at detecting UB, for example.

emarteca replied:
Ok that makes sense -- sounds like a plan.

oli-obk replied:
> One small issue with both of these solutions that I'm currently running into is that the failing test output is now different when a SO is provided and the function being called is not found (since the error will say something like "function not exported in SO").

This is why I would like for that not to be an error, but instead return `Ok(false)` or sth. Then you can choose to ignore it and continue to the shims.

emarteca replied:
Good point -- in the current version of the code I'm checking the arguments [before trying to load the function from the SO file](https://github.com/emarteca/miri/blob/9db25a1b8d9fb31aa908eadebaf6ed87c1647699/src/c_ffi_support.rs#L179). Swapping the order of these computations fixes it.

However, rerunning all the tests with the SO file it looks like some functions seem to be implicitly exported by the SO file, like `malloc`. For example, [this test](https://github.com/emarteca/miri/blob/9db25a1b8d9fb31aa908eadebaf6ed87c1647699/tests/pass/function_calls/disable_abi_check.rs#L16) fails with the message:
```
>error: unsupported operation: Unsupported scalar argument type to external C function: $HEX
>  --> $DIR/disable_abi_check.rs:LL:CC
>   |
>LL |         let _ = malloc(0);
>   |                 ^^^^^^^^^ Unsupported scalar argument type to external C function: $HEX
```
Which is because the function _is_ retrieved from the SO file, but it then expects a hex argument which is not one of the types that's supported right now (although, that reminds me I need to add support for `usize` and `isize`).

oli-obk replied:
Oh fun. Uh... I have no good solution right now beyond silently returning `OK(false)` for unsupported things instead of erroring out. That said, since this is all experimental I'd be fine with falling back to "errors are silently ignored" for now, until there is no more unsupported api surface and all errors are true API mismatches

oli-obk reviewed (changes requested):
I think we'll limit everything to linux for now and then look into expanding the support (and CI coverage) to other platforms in separate PRs. Please set it up so that the shared object is only built on linux for now

RalfJung commented:
It would be great if you could resolve those conflicts by doing a rebase over latest master, so that CI can run on this PR.

oli-obk commented on tests/fail/external_C/function_not_in_SO.stderr:
```diff
@@ -0,0 +1,14 @@
+LL |         foo();
+   |         ^^^^^ can't call foreign function: foo
+   |
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
```
oh, we could extend this diagnostic to also suggest the new ffi flag

emarteca replied:
good idea!

emarteca commented on tests/fail/external_C/bad_args.rs:
```diff
@@ -0,0 +1,12 @@
+// compile-flags: -Zmiri-external_c_so_file=tests/external_C/libtestlib.so
+
+extern "C" {
+    fn printer();
```
Ah you're right -- and it won't error if the signature in `extern "C"` matches the way it is used in the Rust code later, if it doesn't break the C code. I've removed this test for now but I'll add it back when we support more argument types and can actually crash the C by passing in wrong values (which I don't think we can do right now, the ints just auto-cast).

emarteca commented:
> It would be great if you could resolve those conflicts by doing a rebase over latest master, so that CI can run on this PR.

It looks like the linux build is failing because the install is trying to download a new dependency `abort_on_panic` but it's supposed to be in offline mode -- if this is the issue, I'm not sure how to add new dependencies to the CI build. 

Error message from CI build log:
```
./miri install
  Installing miri v0.1.0 (/home/runner/work/miri/miri)
error: failed to compile `miri v0.1.0 (/home/runner/work/miri/miri)`, intermediate artifacts can be found at `/home/runner/work/miri/miri/target`
Caused by:
  failed to download `abort_on_panic v2.0.0`
Caused by:
  attempting to make an HTTP request, but --offline was specified
```

`abort_on_panic` is a dependency of `libffi`: [see in `Cargo.lock`](https://github.com/emarteca/miri/blob/int-function-args-returns/Cargo.lock#L311)

oli-obk commented on src/c_ffi_support.rs:
```diff
@@ -110,6 +121,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     }
+                    hir::PrimTy::Int(IntTy::Isize) => {
+                        let x = call::<isize>(ptr, &libffi_args.as_slice());
+                        if isize::BITS == 64 {
```
I guess we should Add a check to the new -Z flag to make sure it's only used on the host and not for cross compiling, otherwise this line would need to check the target's bitwidth

oli-obk replied:
I thought we had a function for  writing an integer with a specific size... if not, create it and use it here and below

emarteca replied:
Ah you're right -- [`write_int`](https://github.com/emarteca/miri/blob/int-function-args-returns/src/helpers.rs#L153) does take its integer argument and writes it to the size of `dest.layout.size`. The issue with `isize` and `usize` is that they don't satisfy the requirement of `impl Into<i128>` that `write_int` needs for its integer argument. 

I've just changed this to cast the `size`s to `i64` or `u64`, since this coverts both 32- and 64-bit machines.
I _think_ this means we also don't need to check the host vs target bitwidth, but correct me if I'm wrong!

RalfJung replied:
FWIW please don't use `as` casts in Miri, always go through `try_from`/`try_into` (or `from`/`into` if it is infallible).

We should really enable that clippy lint...

oli-obk commented:
> It looks like the linux build is failing because the install is trying to download a new dependency `abort_on_panic` but it's supposed to be in offline mode -- if this is the issue, I'm not sure how to add new dependencies to the CI build.

I think it has been so long since we added a dependency to miri itself, we've since made CI broken in case we do.

@RalfJung I guess we should remove the `--offline` from the install command and require ppl to pass it themselves if needed?

RalfJung commented:
> I guess we should remove the --offline from the install command and require ppl to pass it themselves if needed?

I think I added it because at the time I used `./miri install` a lot and was annoyed that it spent so much time querying the registry for yanked packages. But maybe that got better, so, yeah, feel free to remove the flag.

oli-obk commented:
Our test suite needs a new feature: `//@only-on-host` before this can work. I'll add it first thing tomorrow

emarteca commented:
> Our test suite needs a new feature: `//@only-on-host` before this can work. I'll add it first thing tomorrow

Thank you! :) :pray:

RalfJung commented:
Can you also make the `--offline` change part of that PR? Would be good to leave "global infrastructure changes" out of this already large PR.

bors commented:
:umbrella: The latest upstream changes (presumably #2377) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
OK, after a rebase everything should work now

oli-obk reviewed (changes requested):
Please rebase instead of merge

bors commented:
:umbrella: The latest upstream changes (presumably #2382) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

emarteca commented on tests/panic/external_C/incorrect_SO_file.stderr:
```diff
@@ -7,7 +7,7 @@ note: the compiler unexpectedly panicked. this is a bug.
 note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
 
-note: rustc 1.64.0-nightly (6077b7cda 2022-07-15) running on x86_64-unknown-linux-gnu
+note: rustc 1.64.0-nightly (db4135175 2022-07-16) running on x86_64-unknown-linux-gnu
```
oh, that's nice

RalfJung commented on tests/external_C/test.c:
```diff
@@ -0,0 +1,31 @@
+#include <stdio.h>
```
Could you rename the folder with the flag, i.e., extern-so or something like that?

RalfJung commented on tests/fail/external_C/function_not_in_SO.rs:
```diff
@@ -0,0 +1,13 @@
+//@only-target-linux
+//@only-on-host
+//@compile-flags: -Zmiri-external_c_so_file=tests/external_C/libtestlib.so
```
Please make sure this test also works when Miri is run via `x.py` in the rustc workspace.
I think currently it will not work since you are doing some setup in `./miri`. The setup needs to happen in `tests/compiletest.rs` instead.

RalfJung commented on tests/panic/external_C/incorrect_SO_file.rs:
```diff
@@ -0,0 +1,16 @@
+    unsafe {
+        // calling a function from a shared object file that doesn't exist
+        printer();
+    }
```
Why should this panic?

`tests/panic` is for tests where, if you run that code in real rustc, it panics. I don't think this is such a test.

This seems to test `-Zmiri` argument parsing. I don't think we have any tests for that, and before we do we would first want to switch the argument parser to a more friendly way of reporting errors.

So please just remove this test for now.

RalfJung commented on tests/pass/external_C/print_from_c.rs:
```diff
@@ -0,0 +1,15 @@
+        printer();
+        printer();
+    }
+}
```
Please merge the two "pass" tests into one.
(We tend to have rather large "pass" tests to avoid having too many of them, since there is a non-trivial constant amount of overhead per test.)

RalfJung reviewed (commented):
I am quite excited by this, but do have a bunch of comments. :)

RalfJung commented on src/lib.rs:
```diff
@@ -37,6 +37,7 @@ extern crate rustc_session;
 extern crate rustc_span;
 extern crate rustc_target;
 
+mod c_ffi_support;
```
In line with my other commants, I think this should be called "ffi_support". If someone wants to write those .so files in Rust, they can. :)

RalfJung commented on src/shims/foreign_items.rs:
```diff
@@ -351,13 +356,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        // First deal with any external C functions in linked .so file
+        // (if any SO file is specified).
+        if this.machine.external_c_so_lib.as_ref().is_some() {
+            if let Some(local_id) = def_id.as_local() {
```
This is suspicious, why can we only do this for local declarations?

It probably has to do with you using HIR types when you should be using MIR types.

emarteca replied:
Yes you're right -- using regular MIR types we can use `def_id`!

oli-obk commented on src/shims/foreign_items.rs:
```diff
@@ -351,13 +356,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        // (if any SO file is specified).
+        if this.machine.external_c_so_lib.as_ref().is_some() {
+            if let Some(local_id) = def_id.as_local() {
+                if let Some(extern_c_fct_rep) = ExternalCFuncDeclRep::from_hir_node(
+                    &tcx.hir().get(tcx.hir().local_def_id_to_hir_id(local_id)),
+                    link_name,
+                ) {
```
More specifically, what you can use instead of the HIR is `tcx.fn_sig(def_id)` (which also has `input` and `output` fields, but the types are `Ty`, which you can take apart for the individual types by matching on the result of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.kind

bors commented:
:umbrella: The latest upstream changes (presumably #2390) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

emarteca commented on src/shims/foreign_items.rs:
```diff
@@ -351,13 +356,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    // the external function.
+                    // As discussed: https://github.com/rust-lang/miri/pull/2363#discussion_r922392879
+                    if let Ok(true) =
+                        this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)
```
Oh cool, thanks!

emarteca replied:
Hmm, [the solution I have](https://github.com/rust-lang/miri/pull/2363/commits/e7a6466dfb54f43855bf9ddcde5396fa4947873f#diff-316670adf2dc787f3e0c3fef1e129a5f41304a9afa80ca918d2ac4eb4297ffbaR184) ended up being a bit of a hack the `libloading` crate will automatically load system libraries like `libc`. I'm not convinced this is a better solution than what was there before @RalfJung @oli-obk

RalfJung replied:
Uh, that is odd. Is this intended behavior of `libloading` to provide symbols that are not exposed by the .so file we are loading?

emarteca replied:
Yeah, on linux [`libloading` is based on `dlsym`](https://docs.rs/libloading/0.7.3/src/libloading/os/unix/mod.rs.html#202) and [`dlsym`](https://linux.die.net/man/3/dlsym) looks through the dependency tree of the library if it can't find the symbol in the library itself. 
I don't really like the idea of restricting the build of the shared library to not depend on `libc`. The main solutions I see to this are:
1. leave in this hack solution
2. go back to checking the shims instead of bailing out if there is any error in calling a function from the library
3. check the shims before checking the library

RalfJung replied:
Isn't it possible to have a library that completely hides its dependencies?

> If the symbol is not found, in the specified library or any of the libraries that were automatically loaded by dlopen() when that library was loaded, dlsym() returns NULL.

Wow what a crazy semantics not to have a way to opt-out of that...

oli-obk replied:
I'd say we go with the hack (solution 1). What do you think?

RalfJung replied:
We do have some kind of precedence here -- `#[no_mangle]` functions declared in Rust can be called via `extern` declarations. And there, we currently *first* try the shim, and only if there is no shim we fall back to these `no_mangle` functions. Additionally we raise an error if a `no_mangle` function is declared for a shim (that happens in `check_abi_and_shim_symbol_clash`).

So I wonder if we should do the same here?
- Have a way to check if a function is actually in that .so file, not one of its dependencies -- looks like you found a way to do that.
- Ensure that there is no overlap between the provided shims and the functions exported from the .so file.

Of course this means adding more shims can break .so file users, which also does not seem great...

oli-obk replied:
I think people will actually want the ability to override shims with FFI functions.

Ivwas considering some flags for controlling this, but we can do that when we actually encounter such a situation in practice.

RalfJung replied:
> I think people will actually want the ability to override shims with FFI functions.

I agree. But then would they also want to do that with functions exported from Rust? There we have the check because this can be unsound (when you overwrite a function like `malloc` or `memcpy` with one that has the wrong behavior).

oli-obk replied:
Considering no one has complained yet, maybe this is not an issue?

For FFI I'd expect this to come up because `malloc` and `free` in C code won't call the miri shims, so something allocated in Rust code can't be freed in FFI code unless we always call the FFI allocator functions

RalfJung replied:
Okay, let's document for now that the .so file takes precedence. There should be a warning saying that Miri has its own file descriptor table, so if you overwrite *any* function that works on file descriptors, you have to overwrite *all* of them -- otherwise you are going to be mixing up Miri and host FDs and that's going to be "fun".

emarteca commented:
Thanks for the comments -- I've addressed them all but I'm not a big fan of for [this one](https://github.com/rust-lang/miri/pull/2363#discussion_r924436123). I'll wait to rebase with master again until we agree on what to do with this.

emarteca commented on src/c_ffi_support.rs:
```diff
@@ -0,0 +1,302 @@
+use libffi::{high::call::*, low::CodePtr};
+
+use rustc_ast::{IntTy, UintTy};
+use rustc_hir::{self as hir, FnRetTy, Node, Ty};
```
Good point! I was unaware of `tcx.fn_sig` and this was the way I'd figured out how to get this information on input/output types for a function. It's much cleaner with the MIR types!

RalfJung commented:
It looks like you have committed the .so file, that seems wrong?

RalfJung commented on tests/compiletest.rs:
```diff
@@ -39,6 +40,25 @@ fn run_tests(mode: Mode, path: &str, target: Option<String>) -> Result<()> {
     }
 
+    // If we're on linux, then build the shared object file for testing external C function calls.
+    if env::consts::OS == "linux" {
```
Can't this be a `cfg` test?

emarteca commented:
> It looks like you have committed the .so file, that seems wrong?

Yes, my bad (forgot to update the `.gitignore` with new test paths).

oli-obk commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,283 @@
+            }
+            // Functions with no declared return type (i.e., the default return)
+            // have the output_type `Tuple([])`.
+            if let TyKind::Tuple(t_list) = external_fct_defn.output_type.kind() && t_list.len() == 0{
```
This can now be merged into the match above

oli-obk commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,283 @@
+            // If the return type of a function is a primitive integer type,
+            // then call the function (`ptr`) with arguments `libffi_args`, store the return value as the specified
+            // primitive integer type, and then write this value out to the miri memory as an integer.
+            match external_fct_defn.output_type.kind() {
```
Hmm... can we avoid the entire `fn_sig` business and just use `dest.layout.ty` here?

oli-obk commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,283 @@
+        }
+        let mut libffi_args = Vec::<CArg>::with_capacity(args.len());
+        for (cur_arg, arg_type) in args.iter().zip(external_fct_defn.inputs_types.iter()) {
+            libffi_args.push(Self::scalar_to_carg(this.read_scalar(cur_arg)?, arg_type, this)?);
```
And similarly use `cur_arg.layout.ty` here

oli-obk commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,283 @@
+        }
+
+        // Get the function arguments, and convert them to `libffi`-compatible form.
+        if args.len() != external_fct_defn.inputs_types.len() {
```
Afaict this can never be false

oli-obk reviewed (changes requested):
Sorry for all this back and forth. I think I found another simplification 🙈

emarteca commented:
> Sorry for all this back and forth. I think I found another simplification 🙈

Ah you're right -- I thought tracking the `fn_sig` would let us catch the cases where:
```
extern "C" {
       fn add_one_int(x: i32) -> i32;
}
fn main() {
      unsafe {
            add_one_int(1, 2, 3); // more args than are declared in the signature
            add_one_int(); // less args than are declared in the signature
            add_one_int(1.5); // arg type mismatch
      }
}
```
But I just tested this and this arg length and type mismatch is actually caught by rust itself.
If we don't need to manually distinguish between the signature and the args passed into the function, then this is a nice simplification!

RalfJung commented:
There is a way to get a mismatch between declared signature and args passed -- by transmuting between different function pointer types.

However, we don't have a "ground truth" for what the actual argument count of a function is. The signature might be wrong and the final function type correct. So yeah I don't see a point in even having such a check here.

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,265 @@
+        // This code is a reimplementation of the mechanism for getting `dli_fname` in `libloading`,
+        // from: https://docs.rs/libloading/0.7.3/src/libloading/os/unix/mod.rs.html#411
+        // using the `libc` crate where this interface is public.
+        let mut info = std::mem::MaybeUninit::<libc::Dl_info>::uninit();
+        unsafe {
+            if libc::dladdr(*func.deref() as *const _, info.as_mut_ptr()) != 0 {
+                if std::ffi::CStr::from_ptr(info.assume_init().dli_fname).to_str().unwrap()
+                    != lib_path.to_str().unwrap()
+                {
+                    return Ok(false);
+                }
+            }
+        }
```
Can you wrap this together with the `lib.get` above into a function that does the action of "lookup a name in the .so file and make sure it is really from, that file and not one of its dependencies"?

RalfJung commented on README.md:
```diff
@@ -285,6 +285,12 @@ environment variable. We first document the most relevant and most commonly used
   `-Zmiri-disable-isolation` is also set.
+* `-Zmiri-extern-so-file=<path to a shared object file>` is an experimental flag for providing support
+  for FFI calls.
+  **WARNING**: If an invalid/incorrect SO file is specified, this can cause undefined behaviour in Miri itself!
```
```suggestion
  **WARNING**: If an invalid/incorrect `.so` file is specified, this can cause undefined behaviour in Miri itself!
  And of course, Miri cannot do any checks on the actions taken by the external code.
```

RalfJung commented on README.md:
```diff
@@ -285,6 +285,12 @@ environment variable. We first document the most relevant and most commonly used
+* `-Zmiri-extern-so-file=<path to a shared object file>` is an experimental flag for providing support
+  for FFI calls.
+  **WARNING**: If an invalid/incorrect SO file is specified, this can cause undefined behaviour in Miri itself!
+  This is [work in progress](https://github.com/rust-lang/miri/pull/2363); 
+  current support is for functions with integer arguments/returns.
```
```suggestion
  This is **work in progress**; currently, only integer arguments and return values are
  supported (and no, pointer/integer casts to work around this limitation will not work;
  they will fail horribly).
```

emarteca commented:
Btw I've merged the upstream changes and the build is passing -- if it looks good to you guys, I can rebase again!

oli-obk commented on tests/compiletest.rs:
```diff
@@ -39,6 +39,25 @@ fn run_tests(mode: Mode, path: &str, target: Option<String>) -> Result<()> {
+                "-fPIC",
+                "-Wl,--version-script=tests/extern-so/libcode.version",
+            ])
+            .output()
```
You need to check `status` for success, otherwise this just checks whether `cc` was executed, not whether it succeeded

oli-obk reviewed (commented):
We'll have to see how well it integrates into rustc's test suite

oli-obk commented:
You can either test it yourself against rustc's test suite by checking out the rust repo locally, changing the src/tools/miri submodule to point to your branch and running `./x.py test src/tools/miri` or wait for me to do that on Tuesday or so. If that succeeds, we can merge this from my perspective, so

@rustbot ready for a final code review

RalfJung commented on README.md:
```diff
@@ -285,6 +285,14 @@ environment variable. We first document the most relevant and most commonly used
   harness](https://github.com/rust-lang/miri/issues/1702). This has no effect unless
   `-Zmiri-disable-isolation` is also set.
+* `-Zmiri-extern-so-file=<path to a shared object file>` is an experimental flag for providing support
+  for FFI calls.
```
```suggestion
  for FFI calls. Functions not provided by that file are still executed via the usual Miri shims.
```

RalfJung commented on README.md:
```diff
@@ -285,6 +285,14 @@ environment variable. We first document the most relevant and most commonly used
+* `-Zmiri-extern-so-file=<path to a shared object file>` is an experimental flag for providing support
+  for FFI calls.
+  **WARNING**: If an invalid/incorrect `.so` file is specified, this can cause undefined behaviour in Miri itself!
+  And of course, Miri cannot do any checks on the actions taken by the external code.
```
```suggestion
  And of course, Miri cannot do any checks on the actions taken by the external code.
  Note that Miri has its own handling of file descriptors, so if you want to replace *some* functions
  working on file descriptors, you will have to replace *all* of them, or the two kinds of
  file descriptors will be mixed up.
```

RalfJung commented on README.md:
```diff
@@ -285,6 +285,14 @@ environment variable. We first document the most relevant and most commonly used
   `TERM` environment variable is excluded by default to [speed up the test
   harness](https://github.com/rust-lang/miri/issues/1702). This has no effect unless
   `-Zmiri-disable-isolation` is also set.
+* `-Zmiri-extern-so-file=<path to a shared object file>` is an experimental flag for providing support
```
Given how very experimental this feature is, please move it down to the other section for more advanced and more unsafe flags.

RalfJung commented on src/machine.rs:
```diff
@@ -356,6 +356,9 @@ pub struct Evaluator<'mir, 'tcx> {
     /// The number of blocks that passed since the last progress report.
     pub(crate) since_progress_report: u32,
+
+    /// Handle of the optional C shared object file
```
```suggestion
    /// Handle of the optional shared object file for external functions.
```

RalfJung commented on src/machine.rs:
```diff
@@ -413,6 +416,18 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
+                // Note: it is the user's responsibility to provide a correct SO file.
+                // WATCH OUT: If an invalid/incorrect SO file is specified, this can cause
+                // undefined behaviour in Miri itself!
+                (
```
What is the point of these parentheses? They seem entirely redundant.

emarteca replied:
`external_so_lib` is an `Option<(libloading::Library, std::path::PathBuf)>`; these parens are to make the tuple

RalfJung replied:
Ah, I didn't read the code properly, sorry for this.

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,285 @@
+    /// a form compatible with C (through `libffi` call).
+    /// Then, convert return from the C call into a corresponding form that
+    /// can be stored in Miri internal memory.
+    fn call_and_add_external_c_fct_to_context(
```
What "context" is this referring to?

emarteca replied:
I renamed it to `call_external_c_fct`

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,285 @@
+use libffi::{high::call::*, low::CodePtr};
```
Please don't use glob imports from other libraries. I just lost a few minutes trying to figure out where the `call` and `arg` functions have been defined...

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,285 @@
+            match dest.layout.ty.kind() {
+                // ints
+                TyKind::Int(IntTy::I8) => {
+                    let x = call::<i8>(ptr, libffi_args.as_slice());
```
This is from libffi, right? Please import it in a way that this ends up being more like `ffi::call` or so, making it clear that we are handing off to libffi here.

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,285 @@
+    /// Convert a `CArg` to a `libffi` argument type.
+    pub fn arg_downcast(&'a self) -> libffi::high::Arg<'a> {
+        match self {
+            CArg::Int8(i) => arg(i),
```
This is from libffi, right? Please import it in a way that this ends up being more like `ffi::arg` or so, making it clear that we are handing off to libffi here.

RalfJung commented on tests/compiletest.rs:
```diff
@@ -39,6 +39,28 @@ fn run_tests(mode: Mode, path: &str, target: Option<String>) -> Result<()> {
         flags.push(target.into());
     }
 
+    // If we're on linux, then build the shared object file for testing external C function calls.
```
Please move this entire new part into a separate function.

RalfJung commented on tests/pass/extern-so/call_extern_c_fcts.rs:
```diff
@@ -0,0 +1,48 @@
+        // test function that returns -10 as an unsigned int
+        assert_eq!(get_unsigned_int(), (-10i32) as u32);
+
+        // test void function that prints from C -- call it twice
```
Why call it twice? Not that I mind, but it seems totally random.^^

emarteca replied:
I just wanted to make sure the function has the same output if it's called multiple times -- but it is pretty random, I'll just call it once :-)

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,285 @@
+                return Ok(CArg::Int64(k.to_i64()?));
+            }
+            TyKind::Int(IntTy::Isize) => {
+                return Ok(CArg::ISize(k.to_machine_isize(cx)?.try_into().unwrap()));
```
```suggestion
                // This will fail if host != target, but then the entire FFI thing probably won't work well
                // in that situation.
                return Ok(CArg::ISize(k.to_machine_isize(cx)?.try_into().unwrap()));
```
Same for the usize case.

RalfJung commented on tests/pass/extern-so/call_extern_c_fcts.rs:
```diff
@@ -0,0 +1,48 @@
+//@only-target-linux
+//@only-on-host
```
I am curious, what happens when you run this on a non-host target (on a Linux host)?

RalfJung commented on tests/compiletest.rs:
```diff
@@ -1,13 +1,35 @@
+            "-shared",
+            "-o",
+            "tests/extern-so/libtestlib.so",
+            "tests/extern-so/test.c",
```
We need to get this to work in `x.py` and on Rust CI. One issue there is that the source directory will be read-only. So we need to do this build somewhere in the target dir. There is a CARGO_TARGET_DIR env var we can probably use for that. Use some 'miri-extern-so' subfolder of CARGO_TARGET_DIR.

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,289 @@
+
+impl<'a> CArg {
+    /// Convert a `CArg` to a `libffi` argument type.
+    pub fn arg_downcast(&'a self) -> libffi::high::Arg<'a> {
```
Does this need to be `pub`?

emarteca replied:
no it doesn't, good catch

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,289 @@
+        dest: &PlaceTy<'tcx, Provenance>,
+        args: &[OpTy<'tcx, Provenance>],
+    ) -> InterpResult<'tcx, bool> {
+        // Get the pointer to the function in the shared object file if it exists.
```
It'd probably be a good idea to start with a check here comparing `env!("TARGET")` (which is the host target, i.e., what Miri was built for) and... uh somewhere in `tcx.sess.target` we should be able to get a `TargetTriple` that we can call `str` on to compare this with.

RalfJung replied:
Ah, `tcx.sess.opts.target_triple`  seems to be it.

emarteca replied:
Ah, makes sense -- i added this as an extra check in `emulate_foreign_item_by_name`, before even trying to call the external C function. If they don't match it just continues to try the shims. D'you think it should error instead?

RalfJung replied:
Ah, I see. I think that check should live in this file.

But yeah it'd also make sense to fail loudly rather than silently, since this is probably not the intended behavior.

RalfJung commented on tests/fail/extern-so/function_not_in_SO.rs:
```diff
@@ -0,0 +1,13 @@
+//@only-target-linux
```
Please no UPPER_CASE filenames.

RalfJung commented on tests/fail/extern-so/function_not_in_SO.rs:
```diff
@@ -1,6 +1,6 @@
 //@only-target-linux
 //@only-on-host
-//@compile-flags: -Zmiri-extern-so-file=tests/extern-so/libtestlib.so
+//@compile-flags: -Zmiri-extern-so-file=target/miri-extern-so/libtestlib.so
```
Uh, I don't think this will work in x.py when the target dir is set to somewhere else.

emarteca replied:
Ah good point -- is there a handy regex I can use to get the target dir here, or should I add the flag through `compiletests.rs`?

RalfJung replied:
Hm, I am honestly not sure what the best way is to do this. If you add the flag in `compiletest.rs` it will be added to *all* tests.

emarteca replied:
That's true, hmm. One option is I could try and make a list of the tests that need it and add it only to those in `compiletest.rs`. The nicest option would be if there was a macro I could use in the compile flags, like `$TARGET_DIR` or something, but that might be a pain to implement.

RalfJung replied:
The best option I can think of for now is to have a separate folder of tests, `tests/pass-extern-so`, and have compiletest pass the right flag to all those tests.

emarteca replied:
ah, sounds good!

RalfJung replied:
Now we have two test suites with just one test each... still the best I can think of for now. Let's see if @oli-obk can come up with anything better.

oli-obk replied:
yea that's the best for now. We can migrate to something better later

RalfJung commented:
@rustbot author

emarteca commented:
> We cannot rely on this script, remember that Miri is also built by `x.py`.
> 
> cargo sets the TARGET for build scripts, doesn't it also set them for the actual build? If not we should use Miri's build script to froward the information.

Hmm, so yeah it seems like [TARGET](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts) a variable just set for build scripts. 
I tried re-exporting TARGET in [`build.rs`](https://github.com/emarteca/miri/blob/int-function-args-returns/cargo-miri/build.rs): specifically, I added:
```
let target = env::var("TARGET").unwrap();
println!("cargo:rustc-env=TARGET={}", target);
```
But it seems to still not work: `TARGET` isn't available in the running program. I'm sure I'm missing something -- is there a custom way to set environment variables for miri, or is this not the right build script?

RalfJung commented:
Yeah that is the wrong build script. That one is for cargo-miri, you'll have to add one for Miri itself.

RalfJung commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,297 @@
+        // Check if host target == the session target.
+        // This check needs to happen _after_ we check if the shared object file does not
+        // export the function: we don't want to throw an error if we are just going to use
+        // one of the shims.
```
I disagree. In fact I think the check should be somewhere way earlier, during initiailization. Passing this flag at all for cross-interpretation is an error.

Also `throw_ub` is definitely wrong here. UB is when the Abstract Machine says you did something wrong. This here is just a misconfiguration of the user, similar to setting `-Zmiri-preemption-rate=notafloat`.

emarteca replied:
Ah ok, sounds good.

RalfJung commented on src/machine.rs:
```diff
@@ -410,6 +413,18 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
             preemption_rate: config.preemption_rate,
             report_progress: config.report_progress,
             since_progress_report: 0,
+            external_so_lib: config.external_so_file.as_ref().map(|lib_file_path| {
```
I think this is the right spot to error if host != target.

bors commented:
:umbrella: The latest upstream changes (presumably #2441) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented on src/eval.rs:
```diff
@@ -128,6 +129,9 @@ pub struct MiriConfig {
     /// Whether Stacked Borrows retagging should recurse into fields of datatypes.
     pub retag_fields: bool,
+    /// The location of a shared object file to load when calling external functions
+    /// TODO! consider allowing users to specify paths to multiple SO files, or to a directory
```
We generally use `FIXME` instead of `TODO` as `TODO` blocks merging in the rustc repo. I don't know if this is true for submodules, but let's be consistent

oli-obk commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,287 @@
+                    },
+                _ => {}
+            }
+            // TODO ellen! deal with all the other return types
```
same here

oli-obk commented on src/shims/ffi_support.rs:
```diff
@@ -0,0 +1,287 @@
+
+#[derive(Debug, Clone)]
+/// Enum of supported arguments to external C functions.
+pub enum CArg {
```
Should probably document why this intermediate enum is used instead of directly invoking `ffi::args` and using `libffi::high::Arg` immediately

oli-obk reviewed (changes requested):
should probably squash commits again after addressing these.

I'm checking out your branch locally now to see if it actually works in `x.py`

oli-obk commented:
> I'm checking out your branch locally now to see if it actually works in `x.py`

All the tests pass! So I guess after my nits, rebase to eliminate any merge commits and squash the other commits.

bors commented:
:umbrella: The latest upstream changes (presumably #2448) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 88a78826154870f9778af49d48974fd180e31df8 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/miri) for this repository.

<!-- @bors r=oli-obk 88a78826154870f9778af49d48974fd180e31df8 -->
<!-- homu: {"type":"Approved","sha":"88a78826154870f9778af49d48974fd180e31df8","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/miri"} -->

bors commented:
:hourglass: Testing commit 88a78826154870f9778af49d48974fd180e31df8 with merge 64185014af8ffd81109cab1d1140c18933369071...
<!-- homu: {"type":"BuildStarted","head_sha":"88a78826154870f9778af49d48974fd180e31df8","merge_sha":"64185014af8ffd81109cab1d1140c18933369071"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/8033454140?check_suite_focus=true)
Approved by: oli-obk
Pushing 64185014af8ffd81109cab1d1140c18933369071 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/8033454140?check_suite_focus=true"},"merge_sha":"64185014af8ffd81109cab1d1140c18933369071"} -->

saethlin commented on .gitignore:
```diff
@@ -2,10 +2,18 @@ target
 perf.data
 perf.data.old
 flamegraph.svg
+<<<<<<< HEAD
+=======
+<<<<<<< HEAD
+tests/extern-so/libtestlib.so
+=======
+>>>>>>> master
+>>>>>>> 58ba05a0 (C FFI support for functions with int args and returns)
```
Is this a merge conflict issue?

oli-obk replied:
yea, I created a fix PR for this

emarteca replied:
oh no -- good catch and thanks for fixing it!

RalfJung commented on src/machine.rs:
```diff
@@ -412,6 +416,24 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
             since_progress_report: 0,
+            external_so_lib: config.external_so_file.as_ref().map(|lib_file_path| {
+                // Check if host target == the session target.
+                if option_env!("TARGET") == Some(target_triple) {
```
So we panic if they are equal? How does that make sense?

RalfJung replied:
This is hiding a bug, the comparison doesn't actually work right now.

RalfJung replied:
Fixed by https://github.com/rust-lang/miri/pull/2511

RalfJung commented:
Unless we can quickly find a solution for [this licensing issue](https://rust-lang.zulipchat.com/#narrow/stream/231349-t-core.2Flicensing/topic/CC0.20license.3F), we might have to revert this PR (or disable some of its code, at least) in order to be able to update Miri in rustc. :/

emarteca commented:
> Unless we can quickly find a solution for [this licensing issue](https://rust-lang.zulipchat.com/#narrow/stream/231349-t-core.2Flicensing/topic/CC0.20license.3F), we might have to revert this PR (or disable some of its code, at least) in order to be able to update Miri in rustc. :/

Oh yikes :cry: 
If this turns into a huge problem I guess I can_ try rebuilding this using a different crate than `libffi` (that is the dependency that introduces the dependency that's causing this issue).

saethlin commented:
Have we reached out to `libffi` about this? Maybe there's something easy that can be done on their end?

RalfJung commented:
Yeah I am trying that: https://github.com/tov/libffi-rs/pull/58.

@emarteca what alternatives to libffi are there?

emarteca commented:
> Yeah I am trying that: [tov/libffi-rs#58](https://github.com/tov/libffi-rs/pull/58).
> 
> @emarteca what alternatives to libffi are there?

Hmm, `libffi` is apparently higher level bindings for [`libffi-sys`](https://crates.io/crates/libffi-sys/), which doesn't depend on `abort_on_panic`, so we could potentially try using `libffi-sys` directly

RalfJung commented:
Honestly it seems easier to fork libffi and remove their use of abort_on_panic.

RalfJung commented:
But I hope https://github.com/tov/libffi-rs/pull/58 won't take too long so we can avoid any such drastic steps. :)
